File: 2023/python/solutions/day02.py
from functools import reduce
from itertools import starmap

from lib.python import ProblemInput, ProblemResult, Solution


class Solution02(Solution):
    @staticmethod
    def results() -> list[ProblemResult]:
        return [8, 2476, 2286, 54911]

    @staticmethod
    def solve_part01(data: ProblemInput, is_sample: bool) -> ProblemResult:
        return sum(g for g, reveals in parse(data) if valid(reveals, 12, 13, 14))

    @staticmethod
    def solve_part02(data: ProblemInput, is_sample: bool) -> ProblemResult:
        return sum(power(reveals) for _, reveals in parse(data))


def parse(lines: list[str]) -> list:
    record = []
    for line in lines:
        game, reveals_str = line.split(":")
        game_id = int(game.split()[1])
        reveal_lists = reveals_str.split(";")
        reveals = [r.split() for rl in reveal_lists for r in rl.split(",")]
        reveals_typed = list(starmap(lambda n, c: (int(n), c), reveals))

        record.append((game_id, reveals_typed))

    return record


def valid(reveals, reds, greens, blues):
    for n, color in reveals:
        limit = [reds, greens, blues][["red", "green", "blue"].index(color)]
        if n > limit:
            return False

    return True


def power(reveals):
    setsize = {"red": 0, "green": 0, "blue": 0}
    for n, color in reveals:
        setsize[color] = max(n, setsize[color])

    return reduce(lambda acc, x: acc * x, setsize.values(), 1)

File: 2023/rust/solutions/day02.rs
use aoc_lib::solution::Solution;
use aoc_lib::types::{ProblemInput, ProblemResult};
use itertools::Itertools;
use std::collections::HashMap;

type Reveals = Vec<(u16, String)>;

pub struct Solution02;

impl Solution02 {
    fn parse(&self, input: ProblemInput) -> Vec<(u16, Reveals)> {
        let mut record = vec![];
        for line in input {
            let line_parts = line.split([':', ';', ',']).collect_vec();
            let (game_id_str, reveal_strs) = line_parts.split_first().unwrap();
            let game_id = game_id_str.split_whitespace().nth(1).unwrap().parse::<u16>().unwrap();
            let reveals = reveal_strs
                .iter()
                .map(|&r| r.split_whitespace().take(2).collect_tuple().unwrap())
                .map(|(num, col)| (num.parse::<u16>().unwrap(), String::from(col)))
                .collect::<Vec<_>>();
            record.push((game_id, reveals));
        }
        record
    }

    fn is_valid(&self, reveals: &Reveals, reds: u16, greens: u16, blues: u16) -> bool {
        let colors = HashMap::from([("red", reds), ("green", greens), ("blue", blues)]);
        reveals.iter().all(|(n, color)| n <= colors.get(color as &str).unwrap())
    }

    fn power(&self, reveals: &Reveals) -> i32 {
        let mut setsize = HashMap::from([("red", 0u16), ("green", 0), ("blue", 0)]);
        for (n, color) in reveals {
            setsize.entry(color).and_modify(|val| *val = (*val).max(*n));
        }

        setsize.values().product::<u16>().into()
    }
}

impl Solution for Solution02 {
    fn solve_version01(&self, input: ProblemInput) -> Option<ProblemResult> {
        Some(
            self.parse(input)
                .iter()
                .filter(|(_, reveals)| self.is_valid(reveals, 12, 13, 14))
                .map(|(g, _)| g)
                .sum::<u16>()
                .into(),
        )
    }

    fn solve_version02(&self, input: ProblemInput) -> Option<ProblemResult> {
        Some(
            self.parse(input)
                .iter()
                .map(|(_, reveals)| self.power(reveals))
                .sum::<i32>()
                .into(),
        )
    }
}
